### PR TITLE
Numpy `basic_indexes(shape)` strategy

### LINKED ISSUE #1930: Add Numpy strategy for `basic_indexes(shape)`
*This issue is extracted from #1829, and depends on `slices()` from #1929.*

MyGrad defines a strategy for [`basic_indexes`](https://github.com/rsokl/MyGrad/blob/62912d8bbb1c79ca3a72578a3e68720e4c29c594/tests/custom_strategies/__init__.py#L272), and we would like to bring a simpler version of it into the upstream `hypothesis.extra.numpy`.

Given an array shape, it generates valid indices for an array of that shape which conform to [the basic (i.e. no-copy) indexing specification](https://www.pythonlikeyoumeanit.com/Module3_IntroducingNumpy/BasicIndexing.html#Basic-Indexing). Examples from this strategy shrink from indices that produce high-dimensional views of the input array, down to indices that select a single element from the array.

An initial implementation could work something like:

```python
def basic_indexes(shape):
    # Validation logic for shape here - consider extracting a check_shape helper.
    return tuples(*(slices(s) for s in shape))
```

An optional extension would add an `allow_ellipsis=True` flag, which picks 0-2 dims to replace with `...` - only a few to ensure that the resulting arrays are typically reasonably small compared to the input shape, but may still generate a slice with a length != the number of dimensions to index.

`np.newaxis` (aka `None`) is also supported to add size-1 dimensions in any position, but we would put this behind a `allow_newaxis=False` flag.  As a disabled-by-default additional feature, we would prefer to receive a follow-up pull request to support `newaxis` after the initial version is complete and released to users.

### PR DESCRIPTION
**Pull request by Zac-HD**
It's (un?)surprisingly tricky to get all the subtle things right here: for example, `np.shares_memory` returns `False` for views that are of size zero!  Anyway, thanks to a lot of feedback from Hypothesis I think I've got it all right now :grin: 

Closes #1930.

### CODE REVIEW

File: hypothesis-python/src/hypothesis/extra/numpy.py [hunk 1]
```diff
@@ -871,6 +873,125 @@ def broadcastable_shapes(shape, min_dims=0, max_dims=None, min_side=1, max_side=
...
+    ``...`` (Ellipsis), and :obj:`numpy:numpy.newaxis`; which when used to index a
+    ``shape``-shaped array will produce either a scalar or a shared-memory view.
+
+    * ``shape``: the array shape that will be indexed, as a tuple of integers >= 0.
+      This must be at least two-dimensional for a tuple to be a valid basic index;
+      for one-dimensional arrays use :func:`~hypothesis.strategies.slices` instead.
+    * ``min_dims``: the minimum dimensionality of the resulting view from use of
+      the generated index.  When ``min_dims == 0``, scalars and zero-dimensional
+      arrays are both allowed.
+    * ``max_dims``: the maximum dimensionality of the resulting view.
```
Comment (rsokl) [inline hypothesis-python/src/hypothesis/extra/numpy.py]:
```suggestion
    * ``max_dims``: the maximum dimensionality of the resulting view. The default
       values is ``max(len(shape), min_dims) + 2``.
```

File: hypothesis-python/tests/numpy/test_gen_data.py [hunk 2]
```diff
@@ -818,3 +818,77 @@ def test_advanced_integer_index_can_generate_any_pattern(shape, data):
...
+    "condition",
+    [
+        lambda ix: Ellipsis in ix,
+        lambda ix: Ellipsis not in ix,
+        lambda ix: np.newaxis in ix,
+        lambda ix: np.newaxis not in ix,
+    ],
+)
+def test_test_basic_indices_options(condition):
+    indexers = nps.array_shapes(min_dims=2, max_dims=32).flatmap(
```
Comment (rsokl) [inline hypothesis-python/tests/numpy/test_gen_data.py]:
```suggestion
    indexers = nps.array_shapes(min_dims=0, max_dims=32).flatmap(
```

^ works once you remove the shape-length check

File: hypothesis-python/tests/numpy/test_gen_data.py [hunk 3]
```diff
@@ -818,3 +818,77 @@ def test_advanced_integer_index_can_generate_any_pattern(shape, data):
...
+def test_test_basic_indices_can_generate_long_ellipsis():
+    # Runs of slice(None) - such as [0,:,:,:,0] - can be replaced by e.g. [0,...,0]
+    find_any(
+        nps.basic_indices(shape=(1, 0, 0, 0, 1), allow_ellipsis=True),
+        lambda ix: len(ix) == 3 and ix[1] == Ellipsis,
+    )
+
+
+@given(
+    shape=nps.array_shapes(min_dims=2, max_side=4)
```
Comment (rsokl) [inline hypothesis-python/tests/numpy/test_gen_data.py]:
```suggestion
    shape=nps.array_shapes(min_dims=0, max_side=4)
```

File: hypothesis-python/tests/numpy/test_gen_data.py [hunk 4]
```diff
@@ -818,3 +818,77 @@ def test_advanced_integer_index_can_generate_any_pattern(shape, data):
...
+    | nps.array_shapes(min_dims=2, min_side=0, max_side=10),
```
Comment (rsokl) [inline hypothesis-python/tests/numpy/test_gen_data.py]:
```suggestion
    | nps.array_shapes(min_dims=2, min_side=0, max_side=10),
```

File: hypothesis-python/src/hypothesis/extra/numpy.py [hunk 5]
```diff
@@ -871,6 +873,125 @@ def broadcastable_shapes(shape, min_dims=0, max_dims=None, min_side=1, max_side=
...
+    order_check("dims", 0, min_dims, max_dims)
+    check_argument(
+        max_dims <= 32,
+        "max_dims=%r, but numpy arrays have at most 32 dimensions" % (max_dims,),
+    )
+    check_argument(
+        len(shape) >= 2,
+        "shape=%r must be at least two-dimensional for basic indexing "
+        "to work with `...` or `np.newaxis`." % (shape,),
+    )
```
[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/extra/numpy.py]: Oh wow... I'm not sure how I missed that.  (I think maybe I triggered advanced indexing on something with an early experiment?)

Very very glad I called you in for a review though :heart_eyes:

**Comment by Zac-HD** [comment]
:hand:

**Pull request merged**